File: 06/gemini-3.0-pro/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	maxWidth := 0
	for scanner.Scan() {
		line := scanner.Text()
		lines = append(lines, line)
		if len(line) > maxWidth {
			maxWidth = len(line)
		}
	}

	// Pad lines
	for i := range lines {
		if len(lines[i]) < maxWidth {
			lines[i] = lines[i] + strings.Repeat(" ", maxWidth-len(lines[i]))
		}
	}

	type Block struct {
		start, end int
	}
	var blocks []Block
	startCol := -1

	for col := 0; col < maxWidth; col++ {
		isEmpty := true
		for _, line := range lines {
			if line[col] != ' ' {
				isEmpty = false
				break
			}
		}

		if isEmpty {
			if startCol != -1 {
				blocks = append(blocks, Block{startCol, col - 1})
				startCol = -1
			}
		} else {
			if startCol == -1 {
				startCol = col
			}
		}
	}

	if startCol != -1 {
		blocks = append(blocks, Block{startCol, maxWidth - 1})
	}

	var totalSumP1 int64 = 0
	for _, b := range blocks {
		totalSumP1 += processBlock(lines, b.start, b.end)
	}
	fmt.Printf("Day 06 Part 1: %d\n", totalSumP1)

	var totalSumP2 int64 = 0
	for _, b := range blocks {
		totalSumP2 += processBlockPart2(lines, b.start, b.end)
	}
	fmt.Printf("Day 06 Part 2: %d\n", totalSumP2)
}

func processBlock(lines []string, start, end int) int64 {
	var numbers []int64
	var op string

	for _, line := range lines {
		segment := strings.TrimSpace(line[start : end+1])
		if segment == "" {
			continue
		}

		if segment == "+" || segment == "*" {
			op = segment
		} else {
			num, err := strconv.ParseInt(segment, 10, 64)
			if err == nil {
				numbers = append(numbers, num)
			}
		}
	}

	if op == "+" {
		var sum int64 = 0
		for _, n := range numbers {
			sum += n
		}
		return sum
	} else if op == "*" {
		var prod int64 = 1
		for _, n := range numbers {
			prod *= n
		}
		return prod
	}
	return 0
}

func processBlockPart2(lines []string, start, end int) int64 {
	var numbers []int64
	var op string

	for col := start; col <= end; col++ {
		var chars []byte
		for _, line := range lines {
			if line[col] != ' ' {
				chars = append(chars, line[col])
			}
		}

		if len(chars) == 0 {
			continue
		}

		lastChar := chars[len(chars)-1]
		if lastChar == '+' || lastChar == '*' {
			op = string(lastChar)
			chars = chars[:len(chars)-1]
		}

		if len(chars) > 0 {
			numStr := string(chars)
			num, err := strconv.ParseInt(numStr, 10, 64)
			if err == nil {
				numbers = append(numbers, num)
			}
		}
	}

	if op == "+" {
		var sum int64 = 0
		for _, n := range numbers {
			sum += n
		}
		return sum
	} else if op == "*" {
		var prod int64 = 1
		for _, n := range numbers {
			prod *= n
		}
		return prod
	}
	return 0
}

File: 06/gemini-3.0-pro/rust/src/main.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("input.txt").expect("Failed to read input");
    let lines: Vec<String> = input.lines().map(|s| s.to_string()).collect();

    if lines.is_empty() {
        return;
    }

    let max_width = lines.iter().map(|l| l.len()).max().unwrap_or(0);
    let padded_lines: Vec<String> = lines
        .iter()
        .map(|l| format!("{:width$}", l, width = max_width))
        .collect();

    let mut blocks: Vec<(usize, usize)> = Vec::new();
    let mut start_col: i32 = -1;

    for col in 0..max_width {
        let mut is_empty = true;
        for line in &padded_lines {
            if line.chars().nth(col).unwrap() != ' ' {
                is_empty = false;
                break;
            }
        }

        if is_empty {
            if start_col != -1 {
                blocks.push((start_col as usize, col - 1));
                start_col = -1;
            }
        } else {
            if start_col == -1 {
                start_col = col as i32;
            }
        }
    }

    if start_col != -1 {
        blocks.push((start_col as usize, max_width - 1));
    }

    let mut total_sum_p1: i64 = 0;
    for (start, end) in &blocks {
        total_sum_p1 += process_block(&padded_lines, *start, *end);
    }
    println!("Day 06 Part 1: {}", total_sum_p1);

    let mut total_sum_p2: i64 = 0;
    for (start, end) in &blocks {
        total_sum_p2 += process_block_part2(&padded_lines, *start, *end);
    }
    println!("Day 06 Part 2: {}", total_sum_p2);
}

fn process_block(lines: &[String], start: usize, end: usize) -> i64 {
    let mut numbers: Vec<i64> = Vec::new();
    let mut op = ' ';

    for line in lines {
        let segment = &line[start..=end];
        let trimmed = segment.trim();
        if trimmed.is_empty() {
            continue;
        }

        if trimmed == "+" || trimmed == "*" {
            op = trimmed.chars().next().unwrap();
        } else if let Ok(num) = trimmed.parse::<i64>() {
            numbers.push(num);
        }
    }

    if op == '+' {
        return numbers.iter().sum();
    } else if op == '*' {
        return numbers.iter().product();
    }
    0
}

fn process_block_part2(lines: &[String], start: usize, end: usize) -> i64 {
    let mut numbers: Vec<i64> = Vec::new();
    let mut op: Option<char> = None;

    for col in start..=end {
        let mut chars: Vec<char> = Vec::new();
        for line in lines {
            let c = line.chars().nth(col).unwrap();
            if c != ' ' {
                chars.push(c);
            }
        }

        if chars.is_empty() {
            continue;
        }

        let last_char = *chars.last().unwrap();
        if last_char == '+' || last_char == '*' {
            op = Some(last_char);
            chars.pop();
        }

        if !chars.is_empty() {
            let num_str: String = chars.into_iter().collect();
            if let Ok(num) = num_str.parse::<i64>() {
                numbers.push(num);
            }
        }
    }

    if let Some(o) = op {
        if o == '+' {
            return numbers.iter().sum();
        } else if o == '*' {
            return numbers.iter().product();
        }
    }
    0
}
